### DISCUSSION TITLE
Using Proxies with Multiple Ports

### DISCUSSION THREAD

**Discussion by btc2100k** [encode/httpx#2578] [Q&A]
I have a proxy provider that has provided me x number of proxies, authenticated as follows:

```http://uname:password@hostname:port[10000-10014]```

I have no problem setting this up as follows:

```proxies = {'http://': 'http://uname:password@hostname:port', 'https://': 'http://uname:password@hostname:port'}```

However, my provider limits me on the number of threads on a PORT basis, so I wrote a function called get_proxy_address that selects a random port in a valid range (for my specific case, they are ports 10000 to 10014, giving me 15 possible ports to use)

I am using httpx in an async fashion, like this:

```async with httpx.AsyncClient(proxies=get_proxy_address(), verify=False) as session:```

However, the get_proxy_address() function returns just one of the proxies, since the function is only evaluated once, and only one of them gets used for the entire async session.

I removing proxies= from the instantiation of http.AsyncClient as session and tried adding it to session.get(url, proxies=proxies), but that gives me the following error:

```
response = await session.get(url, headers=headers, proxies=get_proxy_address())
TypeError: get() got an unexpected keyword argument 'proxies'
```

Any ideas on how I could rotate through my 15 ports in a singular httpx.AsyncClient session?

Thank you so much in advance!

**Comment by btc2100k** [comment]
I have also tried:

```session._init_proxy_transport(proxy=get_proxy_address())```

Before each request to no avail

**Comment by lovelydinosaur** [comment]
I would suggest using multiple clients for this use case.
Each client will be connected to a different proxy, and use a separate connection pool.

You could also handle this functionality using a *single client* with a custom transport class that rotates outgoing requests across multiple sub-transports, each configured to a different proxy port. *(Perhaps there's scope for a third party package or example GitHub gist there.)*
